Build waterfall chart with candlestick series and fix bugs

alxnddr opened the pull request:
Closes https://github.com/metabase/metabase/issues/37605
Closes https://github.com/metabase/metabase/issues/36832
Closes https://github.com/metabase/metabase/issues/37321
Closes https://github.com/metabase/metabase/issues/37272
Closes https://github.com/metabase/metabase/issues/37056
Closes https://github.com/metabase/metabase/issues/37050

### Description

Implements the waterfall charts using candlestick series instead of stacked bars.

### How to verify

Checkout test cases below, try reproduce the issues:

### Demo

#### [[dc.js migration] waterfall chart click actions are not correct](https://github.com/metabase/metabase/issues/37605#top)

Disabled click on the Total bar since it is "artificial".

#### [[dc.js migration] missing tick at zero](https://github.com/metabase/metabase/issues/36832#top)

<img width="554" alt="Screenshot 2024-02-13 at 10 51 31 PM" src="https://github.com/metabase/metabase/assets/14301985/c23b015e-d702-42ed-bd2e-af67a02754d6">

#### [[dc.js migration] broken log y-axis scale](https://github.com/metabase/metabase/issues/37050#top)

https://github.com/metabase/metabase/assets/14301985/52393242-7eca-44f6-9f7e-9a818eefa272
(I did not embed the video due to Github bug which hides all the content after an embedded video)

#### [[dc.js migration] "total" label displays as NaN sometimes](https://github.com/metabase/metabase/issues/37056#top)

The same chart now:
<img width="539" alt="Screenshot 2024-02-13 at 10 56 15 PM" src="https://github.com/metabase/metabase/assets/14301985/6f984c07-2f46-4da5-9fe6-44dd88d7aa2d">

#### [[dc.js migration] data label can overlap with x-axis](https://github.com/metabase/metabase/issues/37272#top)

Labels do not overlap with ticks anymore:
<img width="541" alt="Screenshot 2024-02-13 at 11 10 20 PM" src="https://github.com/metabase/metabase/assets/14301985/3cacafdc-0a17-47d7-9a41-82c2bc5a75fa">

#### [[dc.js migration] "total" label is sometimes a date instead](https://github.com/metabase/metabase/issues/37321#top)
I generally improved the Total label appearance on the timeseries charts, for example [this](http://localhost:6006/?path=/story/static-viz-waterfallchart--timeseries-x-scale) and [this](http://localhost:6006/?path=/story/static-viz-waterfallchart--timeseries-x-scale-unsorted) story showed incorrectly showed "January 1st" tick instead of "Total" even though there is no "January 1st" in the dataset.

However, in the GH issue there is a valid screenshot which shows the tick of the value right before the total value and since there is no space for the total tick, it is hidden.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

alxnddr committed:
revamp the echarts implementation of waterfall to fix bugs

alxnddr commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/37947?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* `master`
  * **PR #36489** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37926** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37926?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37908** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37908?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #37947** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37947?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
    * **PR #37903** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37903?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37901** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37901?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37871** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37871?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37847** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37847?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37843** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37843?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37842** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37842?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37835** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37835?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37776** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37776?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37773** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37773?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37772** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37772?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37728** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37721** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37721?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #37743** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37743?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37680** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37680?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37673** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37621** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37621?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #37769** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37769?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37603** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37603?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37543** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37543?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37538** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37538?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37489** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37488** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37488?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37487** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37487?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37486** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37282** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37282?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37277** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37277?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37276** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37276?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37247** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37247?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37196** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37196?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37143** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37143?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37142** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37142?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37094** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37094?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37088** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37088?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37064** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37064?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #37063** <a href="https://app.graphite.dev/github/pr/metabase/metabase/37063?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36961** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36961?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36957** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36957?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36926** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36926?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36925** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36925?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36888** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36888?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36887** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36887?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36886** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36886?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36884** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36884?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36883** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36883?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36882** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36882?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36879** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36879?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36878** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36878?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36781** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36781?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36779** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36779?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36728** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36728?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36621** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36621?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36564** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36564?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36560** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36560?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36548** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36548?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36510** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36510?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36502** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36502?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36486** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36486?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36317** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36317?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36135** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36135?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #36563** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36563?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #36729** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36729?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #36013** <a href="https://app.graphite.dev/github/pr/metabase/metabase/36013?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35997** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35997?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35923** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35922** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35922?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35783** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35783?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #35996** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35996?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35573** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35573?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35395** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35395?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35345** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35345?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35321** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35321?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35307** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35307?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35304** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35304?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35303** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35303?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35301** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35301?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35300** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35300?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35297** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35297?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #35118** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35118?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #34820** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34820?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34923** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34923?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34717** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34717?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #34772** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34772?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
          * **PR #34809** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34809?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
            * **PR #34816** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34816?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
              * **PR #34859** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34859?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                * **PR #34869** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34869?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                  * **PR #34872** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34872?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
                    * **PR #35068** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35068?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #34703** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34703?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #34763** <a href="https://app.graphite.dev/github/pr/metabase/metabase/34763?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/37947?utm_source=stack-comment).

alxnddr committed:
use candlestick series for the waterfall, change the total label logic, define min interval for x-axis ticks, extract event handling code making it reusable

alxnddr committed:
fix relative date x-axis, fix pow, log y-axis

alxnddr committed:
fix waterfall Total tooltip, fix y-axis ticks overlap axis name, fix relative date time units

alxnddr committed:
reorganize waterfall code

alxnddr committed:
cleanup, add spec

alxnddr commented on frontend/src/metabase/visualizations/components/EChartsRenderer/ResponsiveEChartsRenderer.tsx:
```diff
@@ -1,16 +1,42 @@
+    ...echartsRenderedProps
+  }: ResponsiveEChartsRendererProps) => {
+    useEffect(() => {
+      onResize(width, height);
```
To correctly compute option we need to use width and height of the chart itself. However, the parent component which is also wrapped in `ExplicitSize` includes legend and paddings so we need to get dimensions from this component

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/model/axis.ts:
```diff
@@ -457,6 +459,26 @@ export function getYAxesModels(
   };
 }
 
+export const getXAxisEChartsType = (
```
Moved it to the model since logic becomes more complex and will continue for the further x-axis fixes

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/option/axis.ts:
```diff
@@ -165,7 +154,11 @@ export const buildDimensionAxis = (
     },
-  };
+    minInterval:
+      timeSeriesInterval != null
```
This is asking ECharts to show ticks with granularity no smaller than the unit size. However, it does not always respect it.

alxnddr commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts:
```diff
@@ -1,22 +1,67 @@
+import { X_AXIS_DATA_KEY } from "metabase/visualizations/echarts/cartesian/constants/dataset";
+import { getNumberOr } from "metabase/visualizations/lib/settings/row-values";
+
+const replaceZerosForLogScale = (dataset: ChartDataset): ChartDataset => {
```
We cannot have zeros on the log scale, however with dc.js we currently still render charts by replacing zeros. Unfortunately, using Number.MIN_VALUE as replacement does not work well with ECharts because it messes up axis range. So if there are zeros and values above one then we replace zeros with ones. Otherwise, we replace zeros with the smallest value from the dataset

alxnddr commented on frontend/src/metabase/visualizations/visualizations/CartesianChart/CartesianChart.tsx:
```diff
@@ -62,230 +48,21 @@ export function CartesianChart(props: VisualizationProps) {
   const legendItems = useMemo(() => getLegendItems(chartModel), [chartModel]);
   const hasLegend = legendItems.length > 1;
 
-  const openQuestion = useCallback(() => {
```
Extracted to a separate hook which hopefully will make it easy to decouple Combo from Waterfall from Scatter later

EmmadUsmani reviewed (commented):
#### [[dc.js migration] waterfall chart click actions are not correct](https://github.com/metabase/metabase/issues/37605#top)

These are still not working correctly, there are two issues

1. The greater than and less than filter actions are missing
2. The equals filter action uses the dimension name, rather than the metric value

https://github.com/metabase/metabase/assets/37751258/f9ceb6fa-27a2-4fd2-ac67-afd3379dfdfa


Also, I'm not sure if we should disable click actions entirely for the total bar, since we would no longer be able to filter by it

https://github.com/metabase/metabase/assets/37751258/377dc523-1424-46a9-8bc1-12b0fd1f5298

EmmadUsmani commented:
#### [[dc.js migration] missing tick at zero](https://github.com/metabase/metabase/issues/36832#top)

Echarts (first screenshot) has a solid line at 0, whereas the previous implementation had a normal dotted split line

<img width="1552" alt="Screenshot 2024-02-14 at 9 51 53 AM" src="https://github.com/metabase/metabase/assets/37751258/529b026f-1ed2-4805-b020-8b6bf5d710e3">

<img width="1890" alt="Screenshot 2024-02-14 at 9 51 43 AM" src="https://github.com/metabase/metabase/assets/37751258/f1ab39e3-fea4-4533-bd86-30da7e8a4545">

EmmadUsmani commented:
The y-axis label for 0 is "-0" on all power scale charts

<img width="1613" alt="Screenshot 2024-02-14 at 9 56 54 AM" src="https://github.com/metabase/metabase/assets/37751258/dba745ee-da13-415e-b9b1-9a81d5e50887">

EmmadUsmani commented:
#### [[dc.js migration] broken log y-axis scale](https://github.com/metabase/metabase/issues/37050#top)

Negative values on log scale no longer work (they did in dc.js)

https://github.com/metabase/metabase/assets/37751258/b047980b-696d-47e4-b9e8-2411226b807a

EmmadUsmani commented:
The original issue with this chart (NaN label) was fixed but now there are a couple more issues
http://localhost:6006/?path=/story/static-viz-waterfallchart--unaggregated-ordinal-2

<img width="564" alt="Screenshot 2024-02-14 at 10 09 48 AM" src="https://github.com/metabase/metabase/assets/37751258/30094a52-f0ed-4e8b-83b6-e932ad7fca7b">

1. The bars are offset to the right
2. There's a thin green line for 0, before we did not show anything. Showing the line is misleading because without data labels you would mistake it for a non-0, non-null value

<img width="1448" alt="Screenshot 2024-02-14 at 10 13 45 AM" src="https://github.com/metabase/metabase/assets/37751258/fdce22f1-59c8-4ba6-b47f-b959312a5314">

EmmadUsmani commented:
#### [[dc.js migration] "total" label is sometimes missing](https://github.com/metabase/metabase/issues/37055#top)

> ECharts intentionally hides the overlapping ticks, this works on wider charts:

This isn't fixed by this PR so I would not mark this as closed, we still should fix it. It's a related to a larger issue where echarts does not render all category labels. I'll work on fixing this as a part of https://github.com/metabase/metabase/issues/37420 since I will be re-implementing the dimension label interval logic to fix that issue anyways.

alxnddr commented:
> Negative values on log scale no longer work (they did in dc.js)

Oh, seems like another limitation which is technically valid. Since it applies to all log charts with negative values, I created an issue for it https://github.com/metabase/metabase/issues/38780

alxnddr commented:
> Echarts (first screenshot) has a solid line at 0, whereas the previous implementation had a normal dotted split line

This is the behavior of ECharts itself that applies to all cartesian charts. I would not call this a bug since it is just a different way and I'm not sure if this worth investing time in since it will probably require some hacks to simulate split lines by ourselves

alxnddr committed:
fix name of waterfall story

alxnddr committed:
reapply hover cursor fix

alxnddr committed:
fix negative zeros on y-axis power scale

alxnddr committed:
fix click events on waterfall, combo, and scatter charts

alxnddr committed:
pick correct dimension column for click event from data

alxnddr commented:
> These are still not working correctly, there are two issues

Thanks for the explanation, fixed that for all cartesian charts.

alxnddr commented:
> The y-axis label for 0 is "-0" on all power scale charts

Good catch, it was a common problem for all charts that used it — fixed it.

> The original issue with this chart (NaN label) was fixed but now there are a couple more issues

I fixed showing thin green line when the value is zero.
Regarding the bars offset, although the story name is "unaggregated-ordinal-2" it is actually linear so I renamed it. Since all charts that use linear x-axis suffer from it the same way, I think it makes sense to fix the problem in the scope if the linear x-axis bug issue.

alxnddr committed:
snapshots

EmmadUsmani commented:
Can we disable the initial animation? Right now it gets interrupted by mouse events, which makes the performance feel bad. Also we don't have this initial animation when loading bar or other charts.

https://github.com/metabase/metabase/assets/37751258/5a0aefdd-a1c6-4c4e-8d37-61dd21461195

In general I think we should just disable animations for now. They are a good idea but I think we should do them after the migration, there's some kinks that need to be ironed out that just aren't worth our time at the moment. For example, animations don't trigger when initially modifying a saved question.

https://github.com/metabase/metabase/assets/37751258/c710b2e6-499f-4cdf-a078-c48a50f90646

alxnddr committed:
disable initial animation on the waterfall chart

alxnddr commented:
@EmmadUsmani I've disabled the animation

EmmadUsmani commented:
Bars are overflowing on this question: http://localhost:3000/question/251-waterfall-bar-overflow-uk-car-accidents

<img width="1905" alt="bars overflow" src="https://github.com/metabase/metabase/assets/37751258/2ec8ca65-23ba-42ed-9150-1bdb1e4cd4b6">

We have a related issue https://github.com/metabase/metabase/issues/36922 but that issue only encompasses `category` axes right now. We can either add this question to the scope of that issue (if it's a more general problem) or fix it here (if it's specific to waterfall charts).

EmmadUsmani commented:
I thought we decided [here](https://www.notion.so/metabase/Time-series-waterfall-charts-with-relative-units-a4d5346591b24813ac171e73403d1d17?pvs=4#bc5dcab5bc274ce2a9ce09ff2c069e08) to only keep ordinal, linear, and timeseries x-axis scales, but I still see log, power, and histogram in this question: http://localhost:3000/question/136-waterfall-unaggregated-linear-waterfall-example-2

<img width="327" alt="Screenshot 2024-02-15 at 12 11 17 PM" src="https://github.com/metabase/metabase/assets/37751258/a054b4c0-d0c7-40b0-a8d5-5d0edbdf59f0">

alxnddr committed:
fixed waterfall bar widths when data has gaps, improve the total tick logic, address review feedback

alxnddr commented:
@EmmadUsmani I fixed bars width in https://github.com/metabase/metabase/pull/37947/commits/4ff8a0358c692bdc6965fafb1c5ab82a97385f23
Regarding axis options, I will remove them separately since this PR has grown a lot already

alxnddr committed:
improve getTotalTimeSeriesXValue

alxnddr committed:
snapshots

alxnddr commented:
@EmmadUsmani I fixed the wrong bar width on data with gaps. Formatting is broken though, but it is not related to this PR:
<img width="1363" alt="Screenshot 2024-02-15 at 8 04 55 PM" src="https://github.com/metabase/metabase/assets/14301985/2261b496-9dca-43e1-9721-35620125b250">

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/model/dataset.ts:
```diff
@@ -1,26 +1,72 @@
+  }
+
+  if (minNonZeroValue < 0) {
+    throw Error(t`X-axis must not cross 0 when using log scale.`);
```
If we just throw an error the user will see the generic "Something's gone wrong" error which is not informative. Before we used to show a specific messaging saying that crossing 0 was incompatible with the waterfall chart.

<img width="1902" alt="Screenshot 2024-02-16 at 10 18 23 AM" src="https://github.com/metabase/metabase/assets/37751258/7843c136-40e5-46d2-a676-039a31c4cf8a">

<img width="1900" alt="Screenshot 2024-02-16 at 10 18 32 AM" src="https://github.com/metabase/metabase/assets/37751258/30d40ca0-d58e-4c8f-91e7-a8c88af5cf89">

Can we restore this error message? We can either do it in this PR or file an issue to do it later.

alxnddr replied:
I created an issue for this https://github.com/metabase/metabase/issues/38881

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts:
```diff
@@ -1,49 +1,274 @@
+      zlevel: CHART_STYLE.series.zIndex,
+    },
+    {
+      id: "waterfall_labels",
```
Let's extract this to a constant and add a prefix like the null character we used in `constants/dataset.ts` to avoid a potential collision with series names.

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts:
```diff
@@ -1,49 +1,274 @@
+    {
+      id: "waterfall_labels",
+      type: "line",
+      zlevel: CHART_STYLE.series.zIndex + 10,
```
Let's use `z` instead of `zLevel` to be consistent with my recent PR https://github.com/metabase/metabase/pull/38825

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts:
```diff
@@ -1,49 +1,274 @@
+        color: "transparent",
+      },
+      symbolSize: 0,
+      labelLayout: params => {
```
Can we extract this to separate function to avoid making the object super long here?

EmmadUsmani commented on frontend/src/metabase/visualizations/echarts/cartesian/waterfall/option/index.ts:
```diff
@@ -1,49 +1,274 @@
-    dataset,
-    ...getAxes(
-      settings,
+    // TODO: extract common options
```
Should we do this now or in another PR?

If it's another PR please file an issue so it's not forgotten

alxnddr committed:
feedback

alxnddr committed:
fix z indices

Commit:
Update Loki Snapshots
